### PR TITLE
mgr/cephadm: Use mon command to set grafana url in the dashboard

### PR DESCRIPTION
**Pull request by sebastian-philipp**
Fixes: https://tracker.ceph.com/issues/44877
Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by sebastian-philipp** [comment]
QA successful:

* http://pulpito.ceph.com/swagner-2020-04-17_09:07:37-rados-wip-swagner-testing-2020-04-16-1954-distro-basic-smithi/
* http://pulpito.ceph.com/swagner-2020-04-17_09:08:55-rados-wip-swagner-testing-2020-04-16-1954-distro-basic-smithi/

ping @ceph/dashboard

File: src/pybind/mgr/dashboard/settings.py [hunk 1]
```diff
@@ -98,6 +98,10 @@ def __getattr__(cls, attr):
             value = False
         else:
             value = stype(mgr.get_module_option(attr, default))
+
+        if value == default and attr in ['GRAFANA_API_URL']:
+            value = stype(mgr.get_module_option_ex('orchestrator', attr, default))
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/settings.py:103]: Please, don't do this:

* this whole `Settings` class should be deprecated and not extended (it made sense in the early days of dashboard, when `OPTIONS` types and defaults were not properly enforced by ceph-mgr C++ code, but since that changed this has caused a few issues as it doesn't match the behaviour of the C++ counterpart).

* this isn't solving the root issue (FQDN URLs), but just forcing users to manually hack it. If we are using an automated/smart deployment tool, I'd expect that this tool will at least take care for assigning usable URLs to services.

* it creates operational debt: users need to remember modifying almost the same setting in 2 different places (cephadm and dashboard). And, will this be the same for Prometheus/Alertmanager/...?

* it's hacky: breaks module isolation (what happens if the orchestrator module is not running?), modifies a behaviour that is not obvious neither for users nor dashboard devels, ...

Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard/settings.py:103]: Big, thanks for the review!

(shuffled the order of bullet points)
> Please, don't do this:
> 
> * it creates operational debt: users need to remember modifying almost the same setting in 2 different places (cephadm and dashboard). And, will this be the same for Prometheus/Alertmanager/...?

The orchestrator setting is internal to the orchestrator and automatically set by cephadm. Users should not need to know its existence. I only care about the dashboard here.

The idea is to make the dashboard default to cephadm's url, without breaking any existing functionality.

> * this isn't solving the root issue (FQDN URLs), but just forcing users to manually hack it. If we are using an automated/smart deployment tool, I'd expect that this tool will at least take care for assigning usable URLs to services.

Hm, the user might **need** to access Grafana via the fqdn, while cephadm doesn't need to know the fqdn at all. Making the fqdn a requirement for cephadm just to make mgr/dashboard happy seems to be wrong to me. The goal of this PR is to no longer break the dashboard so that users still can overwrite things using the dashboard documentation.

Right now, we're breaking the dashboard by making this setting unusable. I don't want that.

> * it's hacky: breaks module isolation (what happens if the orchestrator module is not running?), modifies a behaviour that is not obvious neither for users nor dashboard devels, ...

Yeah, we need a simple API between mgr/dashboard and the orchestrator. I for one dislike the current approach of defining an api:

https://github.com/ceph/ceph/blob/36ddb9ed3b8a19ffcbd4b71817bc97df9cd07ac9/src/pybind/mgr/orchestrator/_interface.py#L1379-L1381

and #33947

As this requires the introduction of all possible values to `ServiceDescription`. 

> * this whole `Settings` class should be deprecated and not extended (it made sense in the early days of dashboard, when `OPTIONS` types and defaults were not properly enforced by ceph-mgr C++ code, but since that changed this has caused a few issues as it doesn't match the behaviour of the C++ counterpart).

Do you have a better idea?

Comment (epuertat) [reply src/pybind/mgr/dashboard/settings.py:103]: > Right now, we're breaking the dashboard by making this setting unusable. I don't want that.

Neither do I, but I dislike more the idea of creating a wormhole from the dashboard code into the orchestrator just to fix an integration issue.

Part of the disagreement here comes, and please correct me if I'm wrong, from the different views on what's the target deployment scenario:
* You maintain that the need of FQDN is exceptional and hence not a concern for cephadm,
* I think that FQDN should be the rule (versus short hostnames or IP addresses) and therefore a key requirement for cephadm. IMHO if you support FQDN, you can take the hostname/IP, but you cannot easily take the other way around.

Now focusing on the specific fix here: I think it hacks its own design. It mixes explicit and implicit configuration. Choose either: cephadm explicitly sets ceph-dashboard settings, or all dashboard settings are implicitly proxied to orchestrator mirror settings (and can also be manually overridden).

Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard/settings.py:103]: > > Right now, we're breaking the dashboard by making this setting unusable. I don't want that.
> 
> Neither do I, but I dislike more the idea of creating a wormhole from the dashboard code into the orchestrator just to fix an integration issue.
> 
> Part of the disagreement here comes, and please correct me if I'm wrong, from the different views on what's the target deployment scenario:
> 
> * You maintain that the need of FQDN is exceptional and hence not a concern for cephadm,
> * I think that FQDN should be the rule (versus short hostnames or IP addresses) and therefore a key requirement for cephadm. IMHO if you support FQDN, you can take the hostname/IP, but you cannot easily take the other way around.

Hm, I was more looking into the general issue of not breaking the [dasbhaord UX](https://docs.ceph.com/docs/master/mgr/dashboard/#dashboard-grafana), than to find a fix for fqdns specifically. 

> 
> Now focusing on the specific fix here: I think it hacks its own design.

we need some kind of API here. cephadm can provide defaults for a lot of settings:

(no idea why the RGW url was broken apart)
* RGW_API_SCHEME
* RGW_API_HOST
* RGW_API_PORT
* RGW_API_ADMIN_RESOURCE

* RGW_API_ACCESS_KEY
* RGW_API_SECRET_KEY
* RGW_API_USER_ID
* RGW_API_SSL_VERIFY

* GRAFANA_API_URL
* GRAFANA_API_USERNAME
* GRAFANA_API_PASSWORD
* GRAFANA_API_SSL_VERIFY

* GANESHA_CLUSTERS_RADOS_POOL_NAMESPACE

* PROMETHEUS_API_HOST
* ALERTMANAGER_API_HOST

* ISCSI_API_SSL_VERIFICATION

(for haproxy:)
* config-key mgr/dashboard/ssl
* config-key mgr/dashboard/standby_behaviour

I need a way to provide default for all of them. not just GRAFANA_API_URL. 

> It mixes explicit and implicit configuration. Choose either: cephadm explicitly sets ceph-dashboard settings, or all dashboard settings are implicitly proxied to orchestrator mirror settings (and can also be manually overridden).

I'm against explicitly setting all of those mgr/dashboard settings, because then there is no way for users to modify them anymore. Then cephadm has to get everything right and 100% complete for each and every user out there.

Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard/settings.py:103]: also, users should be able to use the dashboard settings to set things in the dashboard. Meaning, the dashboard documentation should continue to work.

Comment (epuertat) [reply src/pybind/mgr/dashboard/settings.py:103]: > I'm against explicitly setting all of those mgr/dashboard settings, because then there is no way for users to modify them anymore. Then cephadm has to get everything right and 100% complete for each and every user out there

> also, users should be able to use the dashboard settings to set things in the dashboard.

What I don't understand here is **why users would (imperatively) want to change a setting directly into the dashboard right after they used a declarative deployment tool** (let that be rook or cephadm). That sounds to me that the deployment tool has some work yet to do.

AFAICS cephadm is so far configuring things in the same way as rook does (and Ceph-Ansible, and DeepSea?): by invoking ceph commands and setting options. That's an explicit push-like API contract. That works for all the settings you enumerated there.

On the other hand, injecting the orchestrator as a man-in-the-middle between ceph-dashboard (or any other ceph-mgr module) and the persistence layer (the config store in this case) sounds really hacky to me. It mixes push and pull configuration approaches, which is weird.

More specific things:
* Many config settings need to trigger side-effects on changes, and they only happen when invoked from commands, so by patching the config store interface, in the best case, will changes go unnoticed to dashboard.
* Additionally, the `Settings` helper does only expose a few config settings, so you will need to globally "patch" `get_module_option()` and *conditionally* re-route all settings to either real dashboard settings or orchestrator clones.
* `get_module_option_ex` does not support localized settings, so you will to implement that new function as well, which will be interesting as it will need to check 4 different config options (a localized setting for dashboard, a general dashboard one, the same localized for orchestrator and finally the global orchestrator one).
* ceph has enough configuration layers (ceph.conf, daemon, config store, ...) so adding a new layer (orchestrator overrides) on top of those... is this really necessary?

Comment (sebastian-philipp) [reply src/pybind/mgr/dashboard/settings.py:103]: > What I don't understand here is **why users would (imperatively) want to change a setting directly into the dashboard right after they used a declarative deployment tool** (let that be rook or cephadm). That sounds to me that the deployment tool has some work yet to do.

I for one can't foresee all the different deployment setups out there. If you really want to break the UX by removing the possibility for users to change dashboard settings (cc @LenzGr ), we should probably restructure the dashboard documentation to make it clear they cannot change dashboard settings when running within a cephadm managed cluster. Wich might actually be a good thing. It's just bold. 

> AFAICS cephadm is so far configuring things in the same way as rook does (and Ceph-Ansible, and DeepSea?):
by invoking ceph commands and setting options. That's an explicit push-like API contract. That works for all the settings you enumerated there.

Sounds sensible. It's then up to the dasbhoard to sort out the fqdn issue. cephadm doesn't need fqdns and should not be forced to do so, just to let dashboard users access grafana.

**Comment by jjahns** [comment]
Hi,

Is there any chance that this will get fixed? We *have* to have the FQDN to access Grafana, as we have certs with a CN containing FQDN, and our users do not have hosts entries on their machines to access the system containing the Grafana container. This is a big deal for operations staff, who have already flagged this as a non-starter.

**Comment by jjahns** [comment]
> Hi,
> 
> Is there any chance that this will get fixed? We _have_ to have the FQDN to access Grafana, as we have certs with a CN containing FQDN, and our users do not have hosts entries on their machines to access the system containing the Grafana container. This is a big deal for operations staff, who have already flagged this as a non-starter.

To add context to this - if you use cephadm --bootstrap, and then FQDN, the grafana_api_url is set to the FQDN, but this is less than ideal, as there are some weird conditions that pop up with the list of services in the dashboard (i.e. any further hosts added only show the mgr daemon in the host list, and mons do not show up).

However, if you don't skip the monitoring components when bootstrapping, it will set the API URL to https://host:3000 (not FQDN).

This needs to be fixed if Grafana is going to be usable in the dashboard.

**Comment by sebastian-philipp** [comment]
@ceph/dashboard I'm agreeing with Ernesto that the cleanest API is the Mon Command API. If that's ok for you, we should use that from now on.

**Comment by jjahns** [comment]
Actually, here is the workaround I utilized on Octopus:
Bootstrap with --skip-monitoring-stack

cephadm shell
ceph orch host set-addr {hostname} {hostname.fqdn}
ceph orch apply grafana {hostname}
ceph orch apply alertmanager {hostname}
ceph orch apply node-exporter {hostname}
ceph orch apply prometheus {hostname}

Tests on our environment result in FQDN being used for GRAFANA_API_URL, which is what we needed in order to have cert complaince, and for clients outside of the environment access to the monitoring dashboards through the Ceph dashboard.

I think the permanent solution needs to be a couple of things:

- Documentation
- Allow cephadm to specify the address of the host similar to orchestrator during bootstrap operations, so the monitoring stack doesn't need to be installed afterwards

File: src/pybind/mgr/cephadm/module.py [hunk 2]
```diff
@@ -1665,15 +1665,49 @@ def _get_hosts(self, label=None):
                 r.append(h)
         return r
 
+    def _verify_add_host(self, spec: HostSpec):
```
Comment (epuertat) [inline src/pybind/mgr/cephadm/module.py:1668]: If we just want to use `spec.hostname`, why not requiring that property alone?

File: src/pybind/mgr/cephadm/module.py [hunk 3]
```diff
@@ -1665,15 +1665,49 @@ def _get_hosts(self, label=None):
...
+            if h.startswith(other + '.') or other.startswith(h + '.'):
+                # TODO: under some artificial circumstances, this makes it impossible to
+                # add hosts, e.g. add_host('myhost') && add_host('myhost.with-a-dot-in-it')
+                raise OrchestratorValidationError(
+                    f"Failed to add {h}: There is already a similar hostname: {other}" + message)
+
+        if self.inventory.keys():
+            fqdns = [i.count('.') > 0 for i in self.inventory.keys()]
+            any_fqds = any(fqdns)
+            if h.count('.') > 0 and not any_fqds:
```
Comment (epuertat) [inline src/pybind/mgr/cephadm/module.py:1689]: I think `any()` and `all()` are pretty understandable:
```suggestion
            if h.count('.') > 0 and not any(fqdns):
```

File: src/pybind/mgr/cephadm/module.py [hunk 4]
```diff
@@ -1665,15 +1665,49 @@ def _get_hosts(self, label=None):
...
+                raise OrchestratorValidationError(
+                    f"Failed to add {h}: Host looks like an FQDN" + message)
+
+            all_fqds = all(fqdns)
+            if h.count('.') == 0 and all_fqds:
```
Comment (epuertat) [inline src/pybind/mgr/cephadm/module.py:1694]: Same here.

File: src/pybind/mgr/cephadm/module.py [hunk 5]
```diff
@@ -1665,15 +1665,49 @@ def _get_hosts(self, label=None):
                 r.append(h)
         return r
 
+    def _verify_add_host(self, spec: HostSpec):
+        """
+        Make sure, we have sane hostname.
+        """
```
Comment (epuertat) [inline src/pybind/mgr/cephadm/module.py:1671]: Maybe it's just me that took me a while to understand this function. Perhaps some comments would help.

So this results in cephadm either allowing all hosts as FQDN or as short hostnames? Is it that?

Wouldn't it be better to make this explicit? For example, FQDN would require '--fqdn' flag, otherwise short hostnames is assumed. I'm not much of a fan of validation rules changing implicitly.

File: src/pybind/mgr/cephadm/module.py [hunk 6]
```diff
@@ -2621,16 +2655,22 @@ def _check_daemons(self):
...
         # make sure the dashboard [does not] references grafana
         try:
-            current_url = self.get_module_option_ex('dashboard',
-                                                    'GRAFANA_API_URL')
+
+            ret, current_url, err = self.mon_command({
+                "prefix": "get-grafana-api-url"
+            }
+
+            )
```
Comment (epuertat) [inline src/pybind/mgr/cephadm/module.py:2663]: Formatting issue?

File: src/pybind/mgr/cephadm/module.py (hunk 5 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/module.py]: > Wouldn't it be better to make this explicit? For example, FQDN would require '--fqdn' flag, otherwise short hostnames is assumed. I'm not much of a fan of validation rules changing implicitly.

that would require 

https://github.com/ceph/ceph/blob/d347db28bcc04c1ee9266338549b5c5c8bd28b05/src/cephadm/cephadm#L2254

to be moved into ceph itself. might be something for a follow-up pr

**Comment by sebastian-philipp** [comment]
```
2020-05-08T14:35:26.215 INFO:ceph.mgr.x.smithi073.stdout:May 08 14:35:25 smithi073 bash[21868]: Traceback (most recent call last):
2020-05-08T14:35:26.215 INFO:ceph.mgr.x.smithi073.stdout:May 08 14:35:25 smithi073 bash[21868]:   File "/usr/share/ceph/mgr/cephadm/module.py", line 1888, in _check_daemons
2020-05-08T14:35:26.215 INFO:ceph.mgr.x.smithi073.stdout:May 08 14:35:25 smithi073 bash[21868]:     ret, current_url, err = self.check_mon_command({"prefix": "get-grafana-api-url"})
2020-05-08T14:35:26.216 INFO:ceph.mgr.x.smithi073.stdout:May 08 14:35:25 smithi073 bash[21868]:   File "/usr/share/ceph/mgr/mgr_module.py", line 1096, in check_mon_command
2020-05-08T14:35:26.216 INFO:ceph.mgr.x.smithi073.stdout:May 08 14:35:25 smithi073 bash[21868]:     raise MonCommandFailed(f'{cmd_dict["prefix"]} failed: {r.stderr}')
2020-05-08T14:35:26.216 INFO:ceph.mgr.x.smithi073.stdout:May 08 14:35:25 smithi073 bash[21868]: mgr_module.MonCommandFailed: get-grafana-api-url failed: command not known
```
http://qa-proxy.ceph.com/teuthology/swagner-2020-05-08_13:49:07-rados-wip-swagner-testing-2020-05-08-1133-distro-basic-smithi/5034678/teuthology.log

**Comment by sebastian-philipp** [comment]
it's up to @ceph/dashboard to allow users to access Grafana.

**Comment by epuertat** [comment]
> it's up to @ceph/dashboard to allow users to access Grafana.

Hey, @sebastian-philipp , why this change of mind?

**Comment by sebastian-philipp** [comment]
Mainly, cause  (1) you took the responsibility to make this work and (2) I don't have any time to work on this.

**Comment by sebastian-philipp** [comment]
Just note the the current workaround of specifying the client's hostname of grafana in cephadm is wrong: cephadm  prefers bare host names. If users need to access grafana using the FQDN, it's the dashboards responsibility to make this possible.

**Comment by epuertat** [comment]
So, this means no FQDN support at all, that including RGW or HAProxy, right?

**Comment by sebastian-philipp** [comment]
> So, this means no FQDN support at all, that including RGW or HAProxy, right?

this means that dashbaord users have different requirements than cephadm.

**Comment by sebastian-philipp** [comment]
e.g.

1. I'm setting up my cluster using hosts in /etc/hosts
2. then I'm getting a signed certificate for my grafana instance using an FQDN
3. and now I need to access grafana in the dasboard, without messing with cephadm.

**Comment by sebastian-philipp** [comment]
> e.g.
> 
> 1. I'm setting up my cluster using hosts in /etc/hosts
> 2. then I'm getting a signed certificate for my grafana instance using an FQDN
> 3. and now I need to access grafana in the dasboard, without messing with cephadm.

Telling users to create a new Ceph cluster, just because they got a certificate work grafana just isn't going to work

**Comment by epuertat** [comment]
> Just note the the current workaround of specifying the client's hostname of grafana in cephadm is wrong: cephadm prefers bare host names. If users need to access grafana using the FQDN, it's the dashboards responsibility to make this possible.

As I sometimes have issues understanding what cephadm is and what is not, I simply tried and replaced cephadm in that sentence with something I know better (ceph-ansible, k8s) and the result doesn't make sense to me. How's it possible that the configuration management tool is not responsible for something as essential as dealing with FQDNs and setting, where appropriate, either the FQDN or the short hostname...

**Comment by sebastian-philipp** [comment]
> As I sometimes have issues understanding what cephadm is and what is not, I simply tried and replaced cephadm in that sentence with something I know better (ceph-ansible, k8s) and the result doesn't make sense to me. How's it possible that the configuration management tool is not responsible for something as essential as dealing with FQDNs and setting, where appropriate, either the FQDN or the short hostname...

Ok, here are the glory details:

Mainly because cephadm has different requirements and demands than ceph-ansible. cephadm has very minimal requirements when it comes to resolving host names etc. Cephadm executes `root@host` where `host` can be resolved in four different ways:

* a custom ssh config resolving the name to an IP
* via an externally maintained `/etc/hosts`
* via `ceph orch host add <hostname> <IP>`
* DNS

Second, Ceph itself uses `hostname` to determine the name of the host and in addition, **cephadm demands that the name of the host given via `ceph orch host add` equals `hostname` on the remote hosts**. Otherwise cephadm can't be sure, the host names returned by `ceph * metadata` match the hosts known to cephadm.

Third, it turns out, there are two **valid** ways to set `hostname`:

1. bare:
  * make `hostname` return the bare host name
  * make `hostname -f` return the FQDN

2. full:
  * make `hostname` return the FQDN
  * make `hostname -s` return the bare host name

quoting `man hostname`:

>   THE FQDN
>       The FQDN (Fully Qualified Domain Name) of the system is the name that the resolver(3) returns for the host name, such as, ursula.example.com.  It is usually the hostname followed by the DNS domain name (the part after the first dot).  You can check
>       the FQDN using hostname --fqdn or the domain name using dnsdomainname.
>
>       You cannot change the FQDN with hostname or dnsdomainname.
>
>       The recommended method of setting the FQDN is to make the hostname be an alias for the fully qualified name using /etc/hosts, DNS, or NIS. For example, if the hostname was "ursula", one might have a line in /etc/hosts which reads
>
>              127.0.1.1    ursula.example.com ursula

Which means, `man hostname` recommends `hostname` to return the bare host name. Which means that Ceph will return the bare hostnames in `ceph * metadata`. Which means that users run `ceph orch host add <bare-name>` to add a host to the cluster. Which means that the dashboard returns the bare host name when pointing the user to Grafana.

Fourth, If the user now get a certificate with the "common name" pointing to the FQDN, we need a place for the user to make the dashboard return the FQDN instead of the bare name used by Ceph. 

To sum up, cephadm uses `hostname` on the machines to identify hosts and the browser uses the DNS name to identify hosts. And both are not necessary identical.
